### DISCUSSION TITLE
Lesson 7: Unit test of start Lottery get not-owner error

### DISCUSSION THREAD

**Discussion by weilainicolas** [smartcontractkit/full-blockchain-solidity-course-py#589] [Q&A]
I am testing it in local environment. This is my test code: 
def test_can_enter():
    # arrange
    if network.show_active() not in LOCAL_BLOCKCHAIN_ENVIRONMENTS:
        pytest.skip()
    lottery = deploy_lottery()
    account = get_account()
    lottery.startLottery({"from": account})
    # act
    lottery.enter({"from": account, "value": lottery.getEntranceFee()})
    # assert
    assert lottery.players(0) == account

I got the following error, when run brownie test 
 lottery.startLottery({"from": account})
E       brownie.exceptions.VirtualMachineError: revert: Ownable: caller is not the owner
Trace step -1, program counter 1091:
E         File "/home/nick/.brownie/packages/OpenZeppelin/openzeppelin-contracts@3.4.0/contracts/access/Ownable.sol", line 43, in Lottery.startLottery:    
E                * @dev Throws if called by any account other than the owner.
E                */
E               modifier onlyOwner() {
E                   require(owner() == _msgSender(), "Ownable: caller is not the owner");
E                   _;
E               }

**Comment by weilainicolas** [comment]
In get_account(), I added a print of the account. 
lottery = deploy_lottery():  deployed by 0x2a9BDb3127A32a825223d40cC35f08B988F714B0
account = get_account(): test_can_enter get account 0x66aB6D9362d4F35596279692F0251Db635165871
Not sure why the account could be different for these 2 lines of code. they are both accounts[0] in the same ganache-cli session.

**Comment by cromewar** [comment]
@weilainicolas  As you posted is complaining about about trying to retrieve from account which is not the owner. The lottery address is not the same as the wallet address, because the contract gets its own when the contract is deployed. If you are actually referring why the account on the lottery is not the same as the new `get_account` because they are on separated instances.

**Reply by weilainicolas** [reply]
Thanks @cromewar for the response. I don't fully understand why there are from separated instance.
When I run the unit test script, brownie opens ganache-cli instance. Both the deployment and calling of the get_account() happens in this instance, and once test is done, ganache-cli turns down. 
Within deployment, get_account() is called as well and returns accounts[0] from ganache-cli. I think I have the same code as Patrick has in the video. Not sure how his code can run without this error. at 7:57:00.

**Reply by cromewar** [reply]
I almost sure es because this line: 
`lottery = deploy_lottery()`
Could you please share your deploy script here?

**Reply by weilainicolas** [reply]
OK, I might find the reason. here is my deploy_lottery() function. I did not define account within this function, rather, I define it outside, and shared by all functions in deploy_lottery.py 

```Python
account=get_account()

def deploy_lottery():
    lottery = Lottery.deploy(
        get_contract("eth_usd_price_feed").address,
        get_contract("vrf_coordinator").address,
        get_contract("link_token").address,
        config["networks"][network.show_active()]["fee"],
        config["networks"][network.show_active()]["keyhash"],
        {"from": account},
        publish_source=config["networks"][network.show_active()].get("verify", False),
    )
    print(f"deployed by {account}")
    return lottery
```

**Reply by cromewar** [reply]
Oh I see now, well as you see the accounts where in different instances. Did you solve the problem with this information?

**Reply by Cosmodude** [reply]
Also met this issue!
Just moved get _account out of functions, defined it before them
